### ISSUE TITLE
`<Modal>` is not accessible by keyboard

### ISSUE THREAD

**Issue by tulup-conner** [themesberg/flowbite-react#647]
**Describe the bug**
`<Modal>`s do not move the focus when opened, so users typically have to tab several times to find a button that can control the modal.

**To Reproduce**

1. Go to https://flowbite-react.com/modal
2. Tab to a Modal and hit Space
3. Tab until your focus reaches one of the modal buttons

**Expected behavior**
The close (X) button, or the primary action button inside the content of the modal should gain focus once the modal is opened.

**Comment by tulup-conner** [comment]
Also, to clarify, you can close modals already via keyboard by hitting Escape, but that isn't necessarily going to be "obvious". The default way keyboard users navigate between areas of focus is by tabbing.

**Comment by ChrisCoastal** [comment]
Hi @tulup-conner, I'd be interested in working on this issue if it is available to someone who has not contributed to the repo before.

To only fix the `close button` gaining focus when the modal is opened issue, an approach could be to set a ref on the [button](https://github.com/themesberg/flowbite-react/blob/8fd9ddcc5fcd8d5926c0e871dd25493d5d2c255f/src/lib/components/Modal/ModalHeader.tsx#L46) and apply focus to it when the modal mounts (though maybe this is more a band-aid solution).

```
  const closeButtonRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    if (closeButtonRef.current) {
      closeButtonRef.current.focus();
    }
  }, []);

  return (
    // ... other JSX
      <button aria-label="Close" type="button" onClick={onClose} ref={closeButtonRef}>
      </button>
)
```

To follow the ADA standard, the complete fix for this issue might involve trapping the focus to items within the modal dialog. The common approach I've seen for this is to find/index all the elements that should be focusable and mimic the native browser behaviour to cycle through these elements on `tab` or  `shift + tab` keydown events.

It may make sense to have `<FocusTrap>` as a component of its own that can be used by `<Modal>` and other components, though it would be a larger project as I see this issue occurs in all the other Flowbite libs.

Happy to hear your thoughts.

**Comment by tulup-conner** [comment]
W3 actually has a specific example for keyboard controls which probably wouldn't hurt to follow: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/

They also recommend a fixed first child element to retain focus. Love the idea and please feel free to ask if you have any problems getting started

**Comment by ChrisCoastal** [comment]
Thanks for the resource! That will be a great guide, I'll give it a look over.
To clarify, are you picturing the fix for this will involve a focus trap or are you just thinking to resolve the initial focus for now?

Thanks, looking forward to it.

**Comment by tulup-conner** [comment]
BTW, beware #568, it may actually be a good idea for you to rebase from that branch to obtain the new composable modal. I don't think your code will be in direct conflict though.

**Comment by ChrisCoastal** [comment]
Hey @tulup-conner, I have what I think is a decent working version of a `<FocusTrap>` component to solve this issue with the Modal (and, I hope, the Dropdown #648 issue as well).

However, I am having trouble getting the current `<Modal>` component itself to work as expected. I can see that it works on [Flowbite React docs page](https://flowbite-react.com/modal), but using the flowbite-react lib I'm unable to reproduce that functionality. As the new `<Modal>` is not yet merged from #568, for the time being I've made a [test sandbox](https://codesandbox.io/p/github/ChrisCoastal/focustrap-test/main?workspace=%257B%2522activeFileId%2522%253Anull%252C%2522openFiles%2522%253A%255B%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clfegprm0000e3b6lso1rq9qj%2522%253A%257B%2522key%2522%253A%2522clfegprm0000e3b6lso1rq9qj%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522clfegv45v00mz3b6ln37voyka%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clfegv35n00ho3b6layzu8u5o%2522%252C%2522isMinimized%2522%253Atrue%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clfegprm0000e3b6lso1rq9qj%2522%252C%2522spacesOrder%2522%253A%255B%2522clfegprm0000e3b6lso1rq9qj%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D) that shows how the component functions and could be used. You can view the proposed [`<FocusTrap>` component here](https://github.com/ChrisCoastal/focustrap-test/blob/21dce972a7c553fc4f7363f5ba2b52a5b457db89/src/components/FocusTrap/FocusTrap.tsx). In my usage, it works with this 'dummy modal' in Chrome, Safari, and Firefox.

A limitation with this `<FocusTrap>` is use with React Portal. It will not work as expected if it has any children using React Portal, which the Flowbite `<Modal>` uses, and so `<FocusTrap>` will have to nested within the `<Modal>` component to work properly. I don't know that there is any easy way around this. However, I think this is a non-issue, as I imagine you will always want the focus functionality in place and having the `<FocusTrap>` as a part of/nested within the `<Modal>` is preferable to the dev using the lib having to wrap `<Modal>` themselves.

Seems best to wait for the PR for #568 to be merged before I look at any changes that may need to be made for the specific use case. As a side note, I think a few of the tests for the `<Modal>` (current and still in the PR for #568) that are testing browser/testing behaviour rather than the `<Modal>` implementation, and could be misleading (for instance [this test](https://github.com/themesberg/flowbite-react/blob/be78e5e748a64ee213018837c576a75ed35a4f13/src/lib/components/Modal/Modal.spec.tsx#L11)).

[Testing for the FocusTrap](https://github.com/ChrisCoastal/focustrap-test/blob/main/src/components/FocusTrap/FocusTrap.spec.tsx) itself presents a challenge as jsdom, which Testing Library uses, does not reproduce in browser `tab` keyboard event behaviour (in my understanding). 
`await user.keyboard('{tab}'); // this will not advance focus in testing`
RTL also doesn't seem to support `ref.current.focus() //  this will not change where the focus is`, so tests are currently failing.
My understanding is that [Puppeteer](https://pptr.dev/) may be a solution for this, but I haven't looked into it as it would be a whole new testing lib to add. There may be a way to write a test with RTL using `createMockNode`.

Let me know what you think of the general approach and if you would like me to work on getting proper testing working. If you simply want to get the initial focus fixed, I think we can just pass the `autoFocus` prop to the `Close Button (X)` as a band-aid solution, and I could open a PR just for that. Also let me know if you would like me to open an issue/PR for the Modal tests.

**Comment by ChrisCoastal** [comment]
Just wanted to check in on this issue and see if there was any desire/direction to move it forward.

**Comment by tulup-conner** [comment]
Hey @ChrisCoastal Thanks for checking it, and sorry for the untimely response. Your PR is greatly appreciated here - I think you're right that what we should do is get #568 out and then tackle this next. An unfortunate side effect of open source, we're all working on things at the same time :) 

Thank you again, and I'll follow up here once that's been merged.

**Comment by ChrisCoastal** [comment]
No apologies needed! It is a pleasure to work on and completely understand. Flowbite has a lot of moving parts to coordinate! I'll keep an eye here and please let me know if there is anything else I can assist with.
